### PR TITLE
MDC context passing over threads

### PR DESCRIPTION
**Pull request by pmohankumar**
Poseidon doesn't pass MDC context between jetty, datasource and filter threads making it impossible to log MDC thread local variables from applications.

Poseidon has to pass MDC context between all the threads that it starts and also has to initialize request context, service context in jetty filter (currently is done in servlet after filters)

### CODE REVIEW

**Comment by pmohankumar** [comment]
Once merged to master, will back-port to 4.x

File: core/src/main/java/com/flipkart/poseidon/legoset/ContextInducedDataSource.java [hunk 1]
```diff
@@ -58,26 +61,38 @@ public DataType call() throws Exception {
...
-            endTrace(dataSource, success);
-            RequestContext.shutDown();
-            ServiceContext.shutDown();
-            HystrixRequestContext.setContextOnCurrentThread(existingState);
-            Brave.getServerSpanThreadBinder().setCurrentSpan(null);
+            shutdownAllContext(existingState, success);
         }
     }
 
+    private void initAllContext() {
```
Comment (shrey-garg) [inline core/src/main/java/com/flipkart/poseidon/legoset/ContextInducedDataSource.java]: can we move this method from here and filter to a Utils or somewhere?

File: core/src/main/java/com/flipkart/poseidon/legoset/ContextInducedDataSource.java [hunk 2]
```diff
@@ -58,26 +61,38 @@ public DataType call() throws Exception {
...
+            MDC.setContextMap(mdcContext);
+        }
+        // Parent thread span info is passed onto filter thread using Brave's ThreadLocal implementation
+        if (serverSpan != null && serverSpan.getSpan() != null) {
+            Brave.getServerSpanThreadBinder().setCurrentSpan(serverSpan);
+        }
+        startTrace(dataSource, request);
+    }
+
+    private void shutdownAllContext(HystrixRequestContext existingState, boolean success) {
```
Comment (shrey-garg) [inline core/src/main/java/com/flipkart/poseidon/legoset/ContextInducedDataSource.java]: this too

File: core/src/main/java/com/flipkart/poseidon/legoset/ContextInducedDataSource.java (hunk 1 above)
Comment (pmohankumar) [inline core/src/main/java/com/flipkart/poseidon/legoset/ContextInducedDataSource.java]: No we don't have a base class for filter & datasource. Remember our earlier discussion of defining Block in lego. Till that's done, this is not doable.

**Comment by pmohankumar** [comment]
@shrey-garg , Introduced ContextInducedBlock as parent for ContextInducedFilter and ContextInducedDataSource

**Pull request merged**